aws_lambda_builders/workflows/nodejs_npm/workflow.py

```diff
@@ -97,48 +68,46 @@ def actions_without_bundler(self, source_dir, artifacts_dir, scratch_dir, manife
 
         npm_copy_npmrc_and_lockfile = NodejsNpmrcAndLockfileCopyAction(tar_package_dir, source_dir, osutils=osutils)
 
-        actions = [
+        self.actions = [
             npm_pack,
             npm_copy_npmrc_and_lockfile,
             CopySourceAction(tar_package_dir, artifacts_dir, excludes=self.EXCLUDED_FILES),
         ]
 
         if self.download_dependencies:
-            # installed the dependencies into artifact folder
-            install_action = NodejsNpmWorkflow.get_install_action(
-                source_dir, artifacts_dir, subprocess_npm, osutils, self.options
+            # install the dependencies into artifact folder
+            self.actions.append(
+                NodejsNpmWorkflow.get_install_action(source_dir, artifacts_dir, subprocess_npm, osutils, self.options)
+            )
+
+        artifacts_cleanup_actions = [
+            NodejsNpmrcCleanUpAction(artifacts_dir, osutils=osutils),
+            NodejsNpmLockFileCleanUpAction(artifacts_dir, osutils=osutils),
+        ]
+
+        # if no dependencies dir, just cleanup artifacts and we're done
+        if not self.dependencies_dir:
+            self.actions += artifacts_cleanup_actions
+            return
+
+        # if we downloaded dependencies, update dependencies_dir
+        if self.download_dependencies:
```

Comment:
are we branching on the same condition twice?

Reply:
One other general comment is we have more than a few branches, we should go down a handler route where we mutate on a a given object all the way through.

eg: self.actions is the one we mutate in this workflow, something like below would be useful to grok through it.


```python
for action_mapper in action_mappers:
    action_mapper.apply(*args, **kwargs)
```

Reply:
> are we branching on the same condition twice?
The first time it doesn't depend on `self.dependencies_dir`. The second time we can assume we have `self.dependencies_dir` (since we checked it before).

Reply:
[L77](https://github.com/aws/aws-lambda-builders/blob/342357c66520a79154cccf5f856c8f763c30f906/aws_lambda_builders/workflows/nodejs_npm/workflow.py#L77) and [L99](https://github.com/aws/aws-lambda-builders/blob/342357c66520a79154cccf5f856c8f763c30f906/aws_lambda_builders/workflows/nodejs_npm/workflow.py#L94)

the conditions are the same, what am I missing?

Reply:
this part in the middle: https://github.com/aws/aws-lambda-builders/blob/342357c66520a79154cccf5f856c8f763c30f906/aws_lambda_builders/workflows/nodejs_npm/workflow.py#L89-L91

Reply:
The first time it doesn't care about, or depend on, `self.dependencies_dir`. The second time, since we have checked `dependencies_dir`, we can assume we have it and we don't have to check for it again. Otherwise it would require some conditional nesting or repetitive conditions, which is what I'm trying to refactor out here 😅

Reply:
to me this is not intuitive and requires a further refactor.

https://github.com/aws/aws-lambda-builders/blob/342357c66520a79154cccf5f856c8f763c30f906/aws_lambda_builders/workflows/nodejs_npm/workflow.py#L109 and https://github.com/aws/aws-lambda-builders/blob/342357c66520a79154cccf5f856c8f763c30f906/aws_lambda_builders/workflows/nodejs_npm/workflow.py#L90 is also repeated.

Reply:
^ the above is also due to something similar. One is performed when there's no dependencies dir and one is performed when there is.

To me the refactor is easier to read than before, but let me know if you disagree. Also open to suggestions on how to further refactor.